---
title: "Class 12 RNA-Seq Analysis"
author: "Carolina Merino (PID: 14484883)"
format: pdf
toc: TRUE
---

## 1. Background


Today we will analyze some RNAseq data from Himes et al. on the effects of a common steroid (dexamethasone) on airway smooth muscle cells (ASM cells)


Our starting point is the "counts" data and "metadata" thatcontain the count values for each gene in their different experiments (i.e. cell lines with or without the drug).


## 2. Bioconductor setup


*Bioconductor package was installed and set up*

```{r}
# Make sure BiocManager is loaded
library(BiocManager)

# Install DESeq2 explicitly in your user library
BiocManager::install("DESeq2", lib = "C:/Users/Linda Kubera/AppData/Local/R/win-library/4.5")

# Load DESeq2
library(DESeq2)

```


## 3. Import countData and colData


```{r}
counts <- read.csv("airway_scaledcounts.csv", row.names=1)
metadata <-  read.csv("airway_metadata.csv")

```


Let's have a wee peek at these objects:


```{r}
head(counts)
head(metadata)
```
> Q1. How many genes are in this dataset? 

There are *38694 genes* in this dataset.

```{r}
nrow(counts)
```
> Q. How many different experiments (columns in counts or rows in metadata) are there?

A total of *8* different experiments are found.

```{r}
ncol(counts)
```
> Q2. How many ‘control’ cell lines do we have?

There are *4* control cell lines found.
```{r}
table(metadata$dex)
```
## 4. Toy differential gene expression


To start our analysis, let's calculate the mean counts for all genes in the "control" experiments. 


1. Extract all "control" columns from the `counts` object.


2. Calculate the mean for all rows (i.e. genes) of these "control" columns


3-4. Do the same for "treated"


5. Compare these `control.mean` and `treated.mean` values.


1 + 2: 

```{r}
control.inds <- metadata$dex == "control"
control.counts <- counts[, control.inds]
control.means <- rowMeans(control.counts)
```

3 + 4:


```{r}
treated.inds <- metadata$dex == "treated"
treated.counts <- counts[, treated.inds]
treated.means <- rowMeans(treated.counts)
```


> Q3. How would you make the above code in either approach more robust? Is there a function that could help here? 

We can use functions like `dplyr::select` or `subset` to avoid relying on column order, and `na.rm = TRUE` in `rowMeans` to handle missing values.



> Q4. Follow the same procedure for the treated samples (i.e. calculate the mean per gene across drug treated samples and assign to a labeled vector called treated.mean)


We calculated the mean per gene across treated samples and stored it in `treated.means`.


```{r}
meancounts <- data.frame(control.means, treated.means)
head(meancounts)
```
Now let's make a plot of control vs. treated mean values for all genes.



>Q5 (a). Create a scatter plot showing the mean of the treated samples against the mean of the control samples. Your plot should look something like the following.

**Answer: See below for the plot:**

```{r}
plot(meancounts)
```

Make this a log log plot


```{r}
plot(meancounts, log="xy")
```

>Q5 (b).You could also use the ggplot2 package to make this figure producing the plot below. What geom_?() function would you use for this plot? 

We would use `geom_point()` function.

```{r}
library(ggplot2)

ggplot(meancounts, aes(x = control.means, y = treated.means)) +
  geom_point()
```


>Q6. Try plotting both axes on a log scale. What is the argument to plot() that allows you to do this?

The argument to the base R function `plot()` that allows you to plot both the x and y axes on a log scale is `log="xy"`.


```{r}
# Make sure the ggplot2 package is loaded
library(ggplot2)

# --- Code for the log-log scatter plot using ggplot2 ---
ggplot(meancounts, aes(x = control.means, y = treated.means)) +
  # Use geom_point() to draw the scatter plot
  geom_point(alpha = 0.5, size = 1.5) +
  # Transform the x-axis to a log10 scale
  scale_x_log10(labels = scales::comma) +
  # Transform the y-axis to a log10 scale
  scale_y_log10(labels = scales::comma) +
  # Set informative labels and title
  labs(
    x = "Control Mean Counts (log10)",
    y = "Treated Mean Counts (log10)",
    title = "Gene Expression: Treated vs. Control (Log-Log Scale)"
  ) +
  # Use a clean theme
  theme_minimal()
```

>Q7. What is the purpose of the arr.ind argument in the which() function call above? Why would we then take the first column of the output and need to call the unique() function?

The argument `arr.ind=TRUE` causes the `which()` function to return a matrix containing the row and column indices of the TRUE values. We use the first column of this matrix because it contains the row indices (genes) of interest, and the `unique()` function ensures that a gene is counted only once, even if it has zero counts across multiple samples (columns).

*Note:*A common threshold used for calling something differentially expressed is a log2(FoldChange) of greater than 2 or less than -2. Let’s filter the dataset both ways to see how many genes are up or down-regulated.

```{r}
up.ind <- meancounts$log2fc > 2
down.ind <- meancounts$log2fc < (-2)
```

>Q8. Using the up.ind vector above can you determine how many up regulated genes we have at the greater than 2 fc level?

Based on the output calculation, the number of up-regulated genes (those with a log2 fold-change greater than 2) is 1846.

```{r}
sum(up.ind, na.rm = TRUE)
```
>Q9. Using the down.ind vector above can you determine how many down regulated genes we have at the greater than 2 fc level?

Based on the output code calculation, the number of down-regulated genes (those with a log2 fold-change less than -2) is 2212.

```{r}
# Assuming 'meancounts' is the correct data frame containing log2fc
down.ind <- meancounts$log2fc < (-2)
sum(down.ind, na.rm = TRUE)
```
>Q10. Do you trust these results? Why or why not?

No, we shouldn’t trust these results yet. Right now, the analysis only looks at how big the change is, but it doesn’t consider how much the results vary between samples. Sometimes a big change can happen just by chance or because of an outlier, so it might not actually be meaningful. To be sure the differences are real, you need to check both the size of the change and whether it’s statistically significant, like using a p-value.

We often talk about metrics like "log2 fold-change"


```{r}
# treated/control
log2(10/40)
```
Let's calculate the log2 fold change for our treated over control mean counts.


```{r}
meancounts$log2fc <- 
  log2(meancounts$treated.means /
  meancounts$control.means)
meancounts$log2fc
```
A common "rule of thumb" is a log2 fold change cutoff of +2 and -2 to call gened "Up regulated" or "Down regulated" respectively. 


Number of Up regulated genes: 


```{r}
sum(meancounts$log2fc >= +2, na.rm=T)
```
Number of Down regulated genes:

```{r}
sum(meancounts$log2fc >= -2, na.rm=T)
```
For the inner nerd: These mean differences might not be as signficant as they look because of outliers in our data that skew the mean.


let's do this analysis properly and keep our inner stats nerd happy. i.e. are the difference we see between drug and no drug sigignificant given the replicate experiments.


```{r, message=FALSE}
library(DESeq2)
```

For DESeq analysis we need three things:


- count values(`countData`)
- metadata telling us about the columns in `countData` (coldata)
- design of experiment (i.e. what do you want to compare?)


Our first function from DESeq2 will setup the input required for analysis by storing all these 3 things together.


```{r}
dds <- DESeqDataSetFromMatrix(countData = counts, 
                              colData = metadata, 
                              design = ~dex)
```
The main function of DESeq that runs the analysis is called `DESeq`.


```{r}
dds <- DESeq(dds)
res <- results(dds)
```
```{r}
head(res)
```


## Volcano Plot

This is a common summary results figure from these types of experiments and plot the log2 fold-change vs the adjusted p-value.


```{r}
plot(res$log2FoldChange, -log(res$padj))
abline(v=c(-2, 2), col="red")
abline(h=-log(0.005), col="red")
```
## Save our results


```{r}
write.csv(res, file="my_results.csv")
```